pmlc/dialect/pxa/transforms/stencil.cc

```diff
@@ -270,19 +509,46 @@ void Stencil::DoStenciling() {
   CollectUsedIndices();
   CollectStrideOneIndices();
 
+  auto ranges = op.getConstantRanges();
+  if (!ranges)
+    return;
+
+  opConstRanges = *ranges;
+
+  // Search tensors' order, inner index and their tiles
+  SearchTensorsOrder();
+
+  IVLOG(1, "Best Perf: " << bestPerf);
+  IVLOG(1, "Best Tiles: " << bestTiles[0] << ":" << bestTiles[1] << ":"
+                          << bestTiles[2]);
+
+  if (bestPerf == std::numeric_limits<double>::max()) {
+    IVLOG(1, "No tile plan for stencil.");
+    return;
+  }
+
   op.setAttr("is_gemm", mlir::UnitAttr::get(op.getContext()));
 }
 
 void StencilPass::runOnFunction() {
   auto func = getFunction();
+  unsigned threads = numThreads.getValue();
   func.walk([&](mlir::AffineParallelOp op) {
-    Stencil as(op);
+    if (threads == 0) {
```

Comment:
Sorry to continue to nitpick, but why is the zero test/replace still inside thee loop?  It only needs to be done once...

Reply:
Yeah, and it will be hoisted out anyway. I don't worry too much about these things. Contemporary compiler have a very good hoisters (especially the SSA based ones). Will move it out, though...

Reply:
I think we should make it an error if the number of threads is not explicitly specified. This is because, if someone writes a `.mlir` test and starts checking that the tiling proceeds in a certain way, then it may work for them due to having say 8 cores, but it will fail on someone else's machine.

Reply:
Personally, I think having a zero-threads == system # of threads is fine, and in production you almost certainly want a way to specify system # of threads.  Regarding hoisting though, this probably would not get hoisted since almost certainly the compiler doesn't know that hardware_concurrency always returns the same number and cannot return zero, esp since it's likely a system call and thus not applicable for inlining.  That said, I'm not actually concerned with perf optimization at a micro level, but just the fact the logically selecting the number of threads is a once per pass decision, so it should look like one.

Reply:
I'm not saying we should never use hardware_concurrency. To be clear I think we should use it in the pipeline definition. But from the pass itself, here, I don't think we should use it. I fear that if someone writes a `.mlir` test and uses the command-line version (via `pmlc-opt`) to access this pass, then they should be forced to set `threads=`.

Reply:
Re: hoisting... This will be hoisted (at least by all 4 C++ compilers I worked on). It operates on a local variable and the compilers assume moving around is fine as long as the variable that is assigned on is not marked volatile. When measured in the past, the optimization is too good to pass it up. And it was widely adopted to allow moving and give a way out with the volatile keyword.

Reply:
No it will *not* be hoisted unless it knows enough details about the hardware_concurrency call, which is non-const (https://en.cppreference.com/w/cpp/thread/thread/hardware_concurrency), and may be internally hit a syscall, or be in a different translation unit.  Imagine that hardware_concurrency returned the value 0 at random 90% of the time, and the value 100 10% of the time (which, unless the compiler can see it's definition, it might).  In that case, it would be reassigned to 100 only around the 10th loop on average and then it would retain it's value and never be called again, since it is inside the if clause.  So clearly it's illegal to hoist it since it would produce invalid behavior to call it only once (it may have side effects, being non-const).  Again, I don't really care about the runtime speed either way, since it's irrelevant in term of overall per, I just wanted it moves up since it's setup logic and not per-loop logic, so it's clearer that way.  Still, if you feel the need to argue the point twice, I can't let it go uncorrected.